File: lib/domain/logic/streak_engine.dart
import '../date_key.dart';
import '../models/models.dart';

/// Freedom Streak + repair tokens (docs/03 §5).
///
/// One streak: consecutive confirmed days at-or-under that day's limit.
/// A repair token absorbs one over-limit day (flame dims, doesn't die);
/// tokens are earned 1 per 7 streak-days, wallet capped at 2.
///
/// Mirrored name-for-name by `functions/src/domain/streakEngine.ts`; any
/// change here lands there too, with parity cases in both test suites.
abstract final class StreakEngine {
  static const int tokenEveryDays = 7;
  static const int tokenWalletCap = 2;

  /// Whether a day keeps the chain alive: confirmed, and either under the
  /// line or saved by a token. The one predicate both walks below share.
  static bool holds(DayLog log) =>
      log.isConfirmed && (!log.isOverLimit || log.repairTokenUsed);

  /// Current streak counted backwards from [today] over [logsByDate].
  /// Days that used a repair token still count.
  static int currentStreak(Map<DateTime, DayLog> logsByDate, DateTime today) {
    var streak = 0;
    var cursor = LpDate.dayStart(today);
    // An unconfirmed (in-progress) or slipped today dims the flame, never
    // kills it (docs/02) — judge the chain from yesterday instead.
    final todayLog = logsByDate[cursor];
    if (todayLog == null || !holds(todayLog)) {
      cursor = LpDate.addDays(cursor, -1);
    }
    while (true) {
      final log = logsByDate[cursor];
      if (log == null || !holds(log)) break;
      streak++;
      cursor = LpDate.addDays(cursor, -1);
    }
    return streak;
  }

  /// The repair-token wallet, derived from history rather than stored.
  ///
  /// Walks every calendar day from the first log through YESTERDAY: each
  /// holding day extends a run, every [tokenEveryDays]th day of a run mints a
  /// token (never past [tokenWalletCap]), and a day that used a token spends
  /// one. A day that does not hold — unlogged, unconfirmed, over the line
  /// with no token — ends the run; the wallet itself carries over, because a
  /// token that was earned and never spent is still earned. Today can only
  /// SPEND: a token is earned by finishing the seventh day, not by starting
  /// it, so the day that would mint a token can never be the day it absorbs.
  /// (Minting mid-day funded a THIRD absorb on day 21 of the QA sim — the
  /// twenty-first holding day minted on its first puff and spent on its
  /// last.)
  ///
  /// Why derived: the wallet used to be stored and re-computed on every
  /// mutation as `streak ~/ 7`, which re-minted a spent token the moment the
  /// next commit ran — over-limit days 15, 20, 21 and 22 of a 30-day sim
  /// were ALL "absorbed" (QA H2, Aug 31 2026). A function of the day map
  /// cannot leak, and the server can recompute the same number, so Ember
  /// never quotes a wallet the Home screen contradicts. Clamped at zero: a
  /// `repairTokenUsed` day the old code let through still holds (nobody
  /// loses a streak over our bug), it just funds nothing.
  static int repairTokens(Map<DateTime, DayLog> logsByDate, DateTime today) {
    if (logsByDate.isEmpty) return 0;
    final todayKey = LpDate.dayStart(today);
    var cursor = logsByDate.keys.reduce((a, b) => a.isBefore(b) ? a : b);
    var run = 0;
    var tokens = 0;
    while (cursor.isBefore(todayKey)) {
      final log = logsByDate[cursor];
      if (log != null && holds(log)) {
        run++;
        if (run % tokenEveryDays == 0 && tokens < tokenWalletCap) tokens++;
        if (log.repairTokenUsed && tokens > 0) tokens--;
      } else {
        run = 0;
      }
      cursor = LpDate.addDays(cursor, 1);
    }
    final todayLog = logsByDate[todayKey];
    if (todayLog != null && todayLog.repairTokenUsed && tokens > 0) tokens--;
    return tokens;
  }

  static FlameState flameFor(int streakDays) =>
      FlameState.forStreak(streakDays);

  /// Whether today's flame renders dimmed (token burned today).
  static bool isDimmed(DayLog? todayLog) => todayLog?.repairTokenUsed ?? false;

  /// Days until the next flame state, or null at Inferno.
  static int? daysToNextFlame(int streakDays) {
    for (final state in FlameState.values) {
      if (state.minDays > streakDays) return state.minDays - streakDays;
    }
    return null;
  }

  /// The next flame state that carries a BADGE — flicker (3), flame (7),
  /// blaze (14), inferno (30) — or null once the inferno is earned. Spark
  /// (1 day) is a flame state but not a milestone, which is how "next" read
  /// "day 0 of 1" after a streak reset (QA L5).
  static FlameState? nextBadgeFlame(int streakDays) {
    for (final state in FlameState.values) {
      if (state == FlameState.spark) continue;
      if (state.minDays > streakDays) return state;
    }
    return null;
  }
}

File: functions/src/domain/streakEngine.ts
/**
 * Port of the streak rules in `lib/domain/logic/streak_engine.dart` +
 * docs/03 §5, for the fields the coach's memory card and the weekly insight
 * need. The client remains the authority on streak *mutation*; this is a
 * read-side recomputation so Ember never quotes a number the app isn't showing.
 */
import {addDays} from './dateKey';
import type {DayLog, FlameState} from './types';

export function flameFor(streakDays: number): FlameState {
  if (streakDays >= 30) return 'inferno';
  if (streakDays >= 14) return 'blaze';
  if (streakDays >= 7) return 'flame';
  if (streakDays >= 3) return 'flicker';
  return 'spark';
}

export const isConfirmed = (log: DayLog): boolean =>
  log.puffs > 0 || log.vapeFreeConfirmed;

/**
 * Whether a day keeps the chain alive. Mirrors `StreakEngine` in
 * `lib/domain/logic/streak_engine.dart`:
 *
 *   holds = isConfirmed && (!isOverLimit || repairTokenUsed)
 *
 * The repair-token clause is load-bearing (docs/03 §5) — a token absorbs one
 * over-limit day so the flame dims rather than dying. Omitting it here made
 * the server disagree with the app about the user's own streak.
 *
 * Exported for `weekStats.ts` — week aggregates must count a token-saved day
 * as on-target exactly the way the flame does.
 */
export const holds = (log: DayLog): boolean =>
  isConfirmed(log) && (log.puffs <= log.limit || log.repairTokenUsed);

/**
 * Consecutive holding days ending today, or yesterday when today does not
 * hold — an in-progress OR slipped today dims the flame instead of zeroing it
 * (pinned by the Dart unit test; don't "fix" it to include today).
 *
 * Anchoring on `holds` rather than `isConfirmed` matters: a confirmed but
 * over-limit today would otherwise be counted, fail the check on the first
 * iteration, and return 0 — erasing the whole streak over a single slip.
 */
export function currentStreak(
  days: Readonly<Record<string, DayLog>>,
  todayKey: string,
): number {
  const today = days[todayKey];
  let cursor = today && holds(today) ? todayKey : addDays(todayKey, -1);
  let streak = 0;
  for (;;) {
    const log = days[cursor];
    if (!log || !holds(log)) return streak;
    streak++;
    cursor = addDays(cursor, -1);
  }
}

/** docs/03 §5: one token per seven holding days, wallet capped at two. */
export const TOKEN_EVERY_DAYS = 7;
export const TOKEN_WALLET_CAP = 2;

/**
 * The repair-token wallet, derived from history. Mirrors
 * `StreakEngine.repairTokens` in `lib/domain/logic/streak_engine.dart`
 * line for line — parity cases in both suites.
 *
 * Walks every calendar day from the first log through YESTERDAY: a holding
 * day extends a run, every seventh day of a run mints (never past the cap),
 * a day that used a token spends one. A non-holding day ends the run; the
 * wallet carries over. Today can only spend — a token is earned by finishing
 * the seventh day, not by starting it. Clamped at zero so a
 * `repairTokenUsed` day the old client leak let through still holds without
 * the wallet going negative.
 *
 * The app used to STORE this and re-derive it as `streak / 7` on every
 * mutation, re-minting spent tokens (QA H2, Aug 31 2026). The memory card
 * reads this rather than `journey.repairTokens` for the same reason it
 * recomputes the streak: Ember must never quote a number the app contradicts.
 */
export function repairTokens(
  days: Readonly<Record<string, DayLog>>,
  todayKey: string,
): number {
  const keys = Object.keys(days);
  if (keys.length === 0) return 0;
  let cursor = keys.reduce((a, b) => (a < b ? a : b));
  let run = 0;
  let tokens = 0;
  while (cursor < todayKey) {
    const log = days[cursor];
    if (log && holds(log)) {
      run++;
      if (run % TOKEN_EVERY_DAYS === 0 && tokens < TOKEN_WALLET_CAP) tokens++;
      if (log.repairTokenUsed && tokens > 0) tokens--;
    } else {
      run = 0;
    }
    cursor = addDays(cursor, 1);
  }
  const today = days[todayKey];
  if (today?.repairTokenUsed === true && tokens > 0) tokens--;
  return tokens;
}

/** Trailing [count] days ending at [todayKey], oldest → newest. */
export function trailingDays(
  days: Readonly<Record<string, DayLog>>,
  todayKey: string,
  count: number,
): DayLog[] {
  const out: DayLog[] = [];
  for (let i = count; i >= 1; i--) {
    const log = days[addDays(todayKey, -i)];
    if (log) out.push(log);
  }
  return out;
}

/**
 * The top-2 puff hours across the window (docs/03 §8). Returns [] until at
 * least three days carry hour data — the app falls back to the onboarding
 * `firstPuff` window before that.
 */
export function dangerHours(window: readonly DayLog[]): number[] {
  const withHours = window.filter((d) => Object.keys(d.hourBuckets).length > 0);
  if (withHours.length < 3) return [];
  const totals = new Map<number, number>();
  for (const day of withHours) {
    for (const [hour, puffs] of Object.entries(day.hourBuckets)) {
      const h = Number.parseInt(hour, 10);
      totals.set(h, (totals.get(h) ?? 0) + puffs);
    }
  }
  return [...totals.entries()]
    .sort((a, b) => b[1] - a[1] || a[0] - b[0])
    .slice(0, 2)
    .map(([hour]) => hour);
}
